### DISCUSSION TITLE
How to handle "Hydration node mismatch"?

### DISCUSSION THREAD

**Discussion by edX3** [nuxt/nuxt#23278] [General]
I'm trying to figure out how to handle Hydration node mismatch. I get loads of those, 40+ per page in a larger website project, it doesn't break anything, everything works, but it bothers me. And I don't know - should it bother me? How do I even begin to debug and solve them? For example one of those is:

> - Client vnode: a 
> - Server rendered DOM: <routerlink class=​"some-class">​</routerlink>​

That's on a regular `<NuxtLink>`. I can't wrap my head around that.
Is there any general guide how to approach this? I would consider myself pretty good in Vue itself, but this stuff is on a whole another level. Please, help!

**Comment by AQian0** [comment]
You can try commenting out the parts of the code that involve v-ifs, refs, and so on that are in general related to the DOM nodes to determine what's wrong.

While this approach is inefficient, it does work.

It's worth noting that this problem usually occurs with nested components and v-ifs. Instead of using the top-level, props-passed data, you could try creating a new judgment condition and assigning a value to it after it's mounted. I think this may involve the order in which the components handle the data...

**Comment by TheAlexLichter** [comment]
I've written an [extensive article](https://www.lichter.io/articles/vue-hydration-error/) about this in the past.

**Reply by AloisSeckar** [reply]
@manniL do you have an idea how to handle mismatch from VueUse's `useNow()`? It renders 1s earlier on the server than on the client:

![image](https://github.com/nuxt/nuxt/assets/7399922/ee14860a-7fb6-4cf2-9be4-6d97df306393)

https://github.com/vueuse/vueuse/discussions/3564

**Reply by TheAlexLichter** [reply]
Either render it only on the client *or* try @danielroe's https://github.com/danielroe/nuxt-time

**Reply by AloisSeckar** [reply]
**Nuxt Time** module looks great, thanks for pointing out

**Comment by baijii** [comment]
Cause neither client-only nor the nuxt-time module was sufficient for my problem, I came up with something that may help other people

```ts
// useSSRNow.ts
import type { UseNowOptions } from '@vueuse/core'

export function useSSRNow(options: UseNowOptions<false> = {}) {
  // use a state to prevent ssr hydration issues
  const serverNow = useState('now', () => useNow())

  // real client-side now that is updating every second
  const {
    now: clientNow,
    isActive,
    resume,
  } = useNow({
    ...options,
    immediate: false,
    controls: true,
  })

  // start the timer on mounted to switch from ssr to csr
  onMounted(() => {
    resume()
  })

  return computed(() => {
    return isActive.value ? clientNow.value : serverNow.value
  })
}
```

```ts
// usage in component
const now = useSSRNow({ interval: 1000 })

const totalSeconds = computed(() => differenceInSeconds(endDate, now.value))

const daysLeft = computed(() => {
  return Math.floor(totalSeconds.value / (24 * 3600))
})

const hoursLeft = computed(() => {
  return Math.floor((totalSeconds.value % (24 * 3600)) / 3600)
})

const minutesLeft = computed(() => {
  return Math.floor((totalSeconds.value % 3600) / 60)
})

const secondsLeft = computed(() => {
  return totalSeconds.value % 60
})

// and so on with very custom render
```
